### PR TITLE
Fix warnings from pydantic

### LINKED ISSUE #3417: Clean up / Fix test warnings resulting from pydantic V2 migration
Created a PR to clean up warnings appearing during  a pytest run:
```
========================================================================= warnings summary ==========================================================================
../../../../Library/Caches/pypoetry/virtualenvs/litellm-vO9t8ogX-py3.11/lib/python3.11/site-packages/pydantic/_internal/_config.py:272: 23 warnings
  /Users/lj/Library/Caches/pypoetry/virtualenvs/litellm-vO9t8ogX-py3.11/lib/python3.11/site-packages/pydantic/_internal/_config.py:272: PydanticDeprecatedSince20: Support for class-based `config` is deprecated, use ConfigDict instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    warnings.warn(DEPRECATION_MESSAGE, DeprecationWarning)

../proxy/_types.py:219
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:219: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../proxy/_types.py:306
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:306: PydanticDeprecatedSince20: `pydantic.config.Extra` is deprecated, use literal values instead (e.g. `extra='allow'`). Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    extra = Extra.allow  # Allow extra fields

../proxy/_types.py:309
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:309: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../proxy/_types.py:338
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:338: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../proxy/_types.py:385
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:385: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../proxy/_types.py:454
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:454: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../proxy/_types.py:466
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:466: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../proxy/_types.py:509
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:509: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../proxy/_types.py:546
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:546: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../proxy/_types.py:840
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:840: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../proxy/_types.py:867
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:867: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../proxy/_types.py:886
  /Users/lj/Documents/litellm/litellm/proxy/_types.py:886: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/
    @root_validator(pre=True)

../utils.py:39
  /Users/lj/Documents/litellm/litellm/utils.py:39: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
====================================================================== short test summary info ======================================================================
ERROR test_azure_perf.py - openai.OpenAIError: Missing credentials. Please pass one of `api_key`, `azure_ad_token`, `azure_ad_token_provider`, or the `AZURE_OPENAI_API_KEY` or `AZURE_OPEN...
ERROR test_config.py - ValueError: api_base is required for Azure OpenAI. Set it on your config. Model - {'model_name': 'gpt-3.5-turbo', 'litellm_params': {'model': 'azure/chatgpt-v-2...
ERROR test_model_alias_map.py - Failed: Error occurred: AnyscaleException - Traceback (most recent call last):
ERROR test_multiple_deployments.py - Failed: An exception occurred: 'NoneType' object has no attribute '_response_ms'
ERROR test_register_model.py - Failed: An error occurred: OpenAIException - Traceback (most recent call last):
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 5 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================== 36 warnings, 5 errors in 2.68s ===================================================================
```

### PR DESCRIPTION
**Pull request by lj-wego**
### before:
```zsh
poetry run pytest . | grep warning
Configuration file exists at /Users/lj/Library/Preferences/pypoetry, reusing this directory.

Consider moving TOML configuration files to /Users/lj/Library/Application Support/pypoetry, as support for the legacy directory will be removed in an upcoming release.
=============================== warnings summary ===============================
../../../../Library/Caches/pypoetry/virtualenvs/litellm-vO9t8ogX-py3.11/lib/python3.11/site-packages/pydantic/_internal/_config.py:284: 25 warnings
    warnings.warn(DEPRECATION_MESSAGE, DeprecationWarning)
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================== 39 warnings, 4 errors in 2.37s ========================
```
### after: 
```zsh
poetry run pytest . | grep warning
Configuration file exists at /Users/lj/Library/Preferences/pypoetry, reusing this directory.

Consider moving TOML configuration files to /Users/lj/Library/Application Support/pypoetry, as support for the legacy directory will be removed in an upcoming release.
=============================== warnings summary ===============================
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================= 1 warning, 4 errors in 2.04s =========================
```

running same test as #3600:
```zsh
poetry run pytest litellm/tests/test_proxy_server.py
Configuration file exists at /Users/lj/Library/Preferences/pypoetry, reusing this directory.

Consider moving TOML configuration files to /Users/lj/Library/Application Support/pypoetry, as support for the legacy directory will be removed in an upcoming release.
================================================================== test session starts ==================================================================
platform darwin -- Python 3.11.3, pytest-7.4.4, pluggy-1.5.0
rootdir: /Users/lj/Documents/litellm
plugins: anyio-4.3.0, asyncio-0.23.6, mock-3.14.0
asyncio: mode=Mode.STRICT
collected 12 items                                                                                                                                      

litellm/tests/test_proxy_server.py s..........s                                                                                                   [100%]

=================================================================== warnings summary ====================================================================
litellm/utils.py:59
  /Users/lj/Documents/litellm/litellm/utils.py:59: DeprecationWarning: open_text is deprecated. Use files() instead. Refer to https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy for migration advice.
    with resources.open_text("litellm.llms.tokenizers", "anthropic_tokenizer.json") as f:

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= 10 passed, 2 skipped, 1 warning in 7.84s ========================================================
```

Not fixing the import lib warning as it is not explicitly related to pydantic.

Regarding pydantic v1 compatibility
https://github.com/BerriAI/litellm/pull/3600#discussion_r1598686881

Openai (1.25.1) is using pydantic v2
https://github.com/openai/openai-python/blob/aed1e43745cd6358b4bafd3a39b3dfeee5e31a03/requirements.lock#L40

## Relevant issues
Fixes #3417 
#3600
#3664
## Type
🐛 Bug Fix
🧹 Refactoring

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -84,3 +84,5 @@ version_files = [
     "pyproject.toml:^version"
 ]
 
+[tool.mypy]
+plugins = "pydantic.mypy"
```
Comment (lj-wego) [inline pyproject.toml:88]: This "fixes" this error:
image

mypy treats `model_config` assignment as an unintended override -- when it is.
Adding the pydantic mypy plugin seems to solve this problem. https://docs.pydantic.dev/latest/integrations/mypy/

After adding this:
mypy . --ignore-missing-imports | grep "Cannot override class variable (previously declared on base class "BaseModel") with instance variable" return zero lint errors relating to overring class variaables.

**Comment by ghost** [comment]
Thanks for this @lj-wego, 

Can you confirm it passes on the list of tests the previous PR attempting to address this failed on - https://github.com/BerriAI/litellm/pull/3664


1. tests/test_ratelimit.py
2. test_organizations.py/test_organization_new
3. Failed linting on types/completion.py, types/router.py, types.embedding.py

**Comment by lj-wego** [comment]
> 1. tests/test_ratelimit.py

image

**Comment by lj-wego** [comment]
> test_organizations.py/test_organization_new

took awhile to figure out what needed to be ran but managed to set up a local docker postgres, prisma migrate and ran the server, and then the test.

image


### results
image

**Comment by lj-wego** [comment]
> Failed linting on types/completion.py, types/router.py, types.embedding.py

i believe this fixes this https://github.com/BerriAI/litellm/pull/3670#discussion_r1604283980

**Comment by lj-wego** [comment]
@krrishdholakia can confirm it works on my machine.

**Comment by rickyyx** [comment]
Thanks! This is really cogging our CI logs. :(

**Comment by lj-wego** [comment]
Merged new changes and updated one pesky `class Config` that was introduced.

Hands tied here for the CI since:
a. i don't have most of the services API keys to properly run the full test/lint suite
b. i don't use circle ci so can't reproduce the testing workflow.

**Comment by ghost** [comment]
Merging in - i'll let you know if there's any issues on ci/cd @lj-wego 


Thank you again for your help here!

**Comment by lj-wego** [comment]
@krrishdholakia 
i notice the CI failing and this might be the issue. 

installs older pydantic
image

image


https://github.com/BerriAI/litellm/blob/deb87f71e330dcaf56574a477d9457d589754cd3/.circleci/requirements.txt#L10

**Comment by lj-wego** [comment]
Added a PR [here](https://github.com/BerriAI/litellm/pull/3938#issue-2327471604) to address this. Unless this is reverted 😓

**Pull request merged**